app/src/main/java/com/example/myweatherapp/viewmodel/MainViewModel.kt

```diff
@@ -0,0 +1,157 @@
+package com.example.myweatherapp.viewmodel
+
+import android.app.Application
+import android.content.Context
+import android.content.SharedPreferences
+import android.util.Log
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.MutableLiveData
+import com.example.myweatherapp.interfaces.IListItem
+import com.example.myweatherapp.models.NewsModel
+import com.example.myweatherapp.repository.NewsRepository
+
+
+class MainViewModel(app: Application) : AndroidViewModel(app) {
+
+
+    private val repository = NewsRepository()
+    private var sharedPreferences: SharedPreferences = app.getSharedPreferences("Url", Context.MODE_PRIVATE)
+
+    private val progressVisibilityLiveData = MutableLiveData<Boolean>()
+    internal val getProgressVisibility = progressVisibilityLiveData
+
+    private val newsData = MutableLiveData<List<IListItem>>()
+    internal val getNewsData = newsData
+
+    private val sourceData = MutableLiveData<List<IListItem>>()
+    internal val getSourceData = sourceData
+
+    private val favoriteData = MutableLiveData<List<IListItem>>()
+    internal val getFavoriteData = favoriteData
+
+    private var spinnerItemsMutableLiveData = MutableLiveData<Set<String>>()
+    internal var getSpinnerItems = spinnerItemsMutableLiveData
+
+    fun loadSpinnerData() {
+        val spinnerList = HashSet<String>()
+        spinnerList.add("all news")
+        return repository.getNews("news") { list ->
+            if (list!=null){
+                list.forEach {
+                    if (it.newsSource.name.isNotEmpty()) {
+                        spinnerList.add(it.newsSource.name)
+                        spinnerItemsMutableLiveData.postValue(spinnerList)
+                    }
+                }
+                progressVisibilityLiveData.postValue(false)
+            }
+        }
+    }
+
+    fun loadNewsData() {
+        progressVisibilityLiveData.postValue(true)
+        return repository.getNews("news") { list ->
+            if (list != null) {
+                newsData.postValue(list)
+                applyIsFavoriteState(list)
+                Log.d("TAG", "GETTING NEWS DATA")
+            } else {
+                Log.d("TAG", "SOME ERROR")
+                progressVisibilityLiveData.postValue(false)
+            }
+        }
+    }
+
+    fun loadSourceData() {
+        progressVisibilityLiveData.postValue(true)
+        return repository.getSource { list ->
+            if (list != null) {
+                newsData.postValue(list)
+                Log.d("TAG", "GETTING SOURCE DATA")
+                progressVisibilityLiveData.postValue(false)
+            } else {
+                Log.d("TAG", "Some ERROR")
```

Comment:
Can't we retrieve more detailed error here? "Some ERROR" does say nothing ;)

Reply:
description of errors has become more detailed